daemon: switch to semaphore-gated WaitGroup for startup tasks

cyphar opened the pull request:
Many startup tasks have to run for each container, and thus using a
WaitGroup (which doesn't have a limit to the number of parallel tasks)
can result in Docker exceeding the NOFILE limit quite trivially. A more
optimal solution is to have a parallelism limit.

In addition, several startup tasks were not parallelised previously
which resulted in very long startup times. According to my testing, 20K
dead containers resulted in ~6 minute startup times (during which time
Docker is completely unusable).

This patch fixes both issues, and the parallelStartupTimes factor chosen
(128 * NumCPU) is based on my own significant testing of the 20K
container case. This patch (on my machines) reduces the startup time
from 6 minutes to less than a minute (ideally this could be further
reduced by removing the need to scan all dead containers on startup --
but that's beyond the scope of this patchset).

In order to avoid the NOFILE limit problem, we also detect this
on-startup and if NOFILE < 2*128*NumCPU we will reduce the parallelism
factor to avoid hitting NOFILE limits (but also emit a warning since
this is almost certainly a mis-configuration).

![benchmarks](https://user-images.githubusercontent.com/2888411/49456934-04fab480-f83e-11e8-969d-69f85873bad4.png)

###### Benchmarks were done on an 8-core machine.

[![cat by Danny VB](https://farm4.staticflickr.com/3488/4051998735_5b4863ac11.jpg)](https://flic.kr/p/7b4y4a)

###### [cat](https://flic.kr/p/7b4y4a) by [Danny VB](https://www.flickr.com/photos/parismadrid/)

Signed-off-by: Aleksa Sarai <asarai@suse.de>

BlaineEXE commented:
I would disagree that there is little benefit to having more start tasks than cores. Goroutines are really lightweight, and having multiple per core allows the scheduler to run tasks while other routines are waiting on IO, and there will be a lot of IO waiting while reading/writing to disk.

I think this could be a good fix if it needs to be released quickly, but it seems much better to me to set the semaphore to limit to some percentage of the process's NOFILE limit.

It also occurs to me that there are a lot of servers with a huge number of cores. I have seen some absurd ones with as many as 1024 cores, and the NOFILE limit on my dev machine is 1024. I would be concerned that it would be possible in some corner cases that limiting to the number of procs might still not prevent the startup process from exceeding the ulimit.

cyphar commented:
My original point about cores was a little flippant -- however, there is a diminishing returns aspect when it comes to goroutines, especially if they are doing very little actual work. Note that "blocking on IO" usually means that the kernel is doing work, and there are many layers of caching that mean us writing to two or three JSON files really is not going to be blocking for any significant period of time. Not to mention that spawning 10K goroutines will result in slower execution because of the significant amount of time spent context-switching (and the number of threads that Go will create each time that one of the 10K goroutines spawns a syscall).

All of that being said, there probably is still a benefit to having (say) 20 goroutines per core or something similar. But it should be noted that the default `NOFILE` is set to `LimitNOFILE=1048576` in `contrib/init/systemd/docker.service` -- so even if we set it to `50%` of `NOFILE` we're still talking about 512K goroutines at worst.

To be clear, I don't really care what the number is (so long as it's reasonable), but "as many containers as we have on the machine" isn't a good number.

BlaineEXE commented:
> To be clear, I don't really care what the number is (so long as it's reasonable), but "as many containers as we have on the machine" isn't a good number.

Definitely agreed on this point. I can't really suggest an educated number for the right number of routines-per-core. Go conventional wisdom I've read suggests having even a few hundred routines per core might be alright, but the startup calls are probably not very lightweight, which kind of throws that out the window. I think the bottom-line of my concern is that this risks increasing the startup time quite a bit.

I would think that doing some reasonably quick benchmarks to see how long the initialization takes with various routines-per-core values would suggest a practical number beyond which things slow down. I think it makes sense that that practical value would be about the same order of magnitude for most systems, so no need to try to do too much optimization. And for machines with a huge number of containers, this might decrease the overall startup time. =)

IMO, Arbitrarily choosing to test startup time by increasing routines-per-core by powers of two seems as reasonable a thing as going by tens or hundreds.

cyphar commented:
Sure, I can do some quick benchmarks (because dead containers are processed on startup this is even simpler).

But (in a slightly snarky done) it should be noted that the bug reported to me had 10K containers and the startup time was taking several minutes (so long that systemd killed Docker because it thought it was timing out). I'm not sure it'd be easy for us to do much worse than that. :wink:

cyphar commented:
After some benchmarking, the actual bottleneck (with 20K dead containers docker takes >6min to start) is not in the parts that currently use goroutines. It's the parts of `restore` which don't (most notable the daemon loading code). The goroutine'd sections take up only about 20s of the 6 minute long operation.

But after some more testing, it does look like there is a difference (though it's not exceptionally significant) between different values of `parallelStartupJobs`. I've also discovered that I made a mistake with the semantics of my Semaphore's `Wait` implementation -- I'm fixing that too.

cyphar commented:
Here's the results of my benchmarks. This was all done on an 8-core machine (with 5 samples of each multiplier -- average plotted with standard-deviation error bars). Looks like the ideal multiplier is `128`.

![performance](https://user-images.githubusercontent.com/2888411/49456934-04fab480-f83e-11e8-969d-69f85873bad4.png)

cyphar committed:
daemon: switch to semaphore-gated WaitGroup for startup tasks

Many startup tasks have to run for each container, and thus using a
WaitGroup (which doesn't have a limit to the number of parallel tasks)
can result in Docker exceeding the NOFILE limit quite trivially. A more
optimal solution is to have a parallelism limit by using a semaphore.

In addition, several startup tasks were not parallelised previously
which resulted in very long startup times. According to my testing, 20K
dead containers resulted in ~6 minute startup times (during which time
Docker is completely unusable).

This patch fixes both issues, and the parallelStartupTimes factor chosen
(128 * NumCPU) is based on my own significant testing of the 20K
container case. This patch (on my machines) reduces the startup time
from 6 minutes to less than a minute (ideally this could be further
reduced by removing the need to scan all dead containers on startup --
but that's beyond the scope of this patchset).

In order to avoid the NOFILE limit problem, we also detect this
on-startup and if NOFILE < 2*128*NumCPU we will reduce the parallelism
factor to avoid hitting NOFILE limits (but also emit a warning since
this is almost certainly a mis-configuration).

Signed-off-by: Aleksa Sarai <asarai@suse.de>

cpuguy83 commented:
Wow, very nice work. I agree we should limit the number of goroutines here.

cpuguy83 commented on daemon/daemon.go:
```diff
@@ -207,68 +216,88 @@ func (daemon *Daemon) restore() error {
 		return err
 	}
 
+	loadWg := syncext.NewLimitedWaitGroup(parallelStartupJobs)
```
If you don't mind, I'd, personally, prefer not to add the new package for this limited case, especially since I feel like it's weird to mix the two things.

cyphar replied:
Would you prefer that I use `WaitGroup` and `chan struct{}` separately, or that I make `LimitedWaitGroup` a new type *just* for `daemon/daemon.go`?

cpuguy83 replied:
Separately 🙏👼

cyphar replied:
Sure, though note that this will change

```go
	loadWg := syncext.NewLimitedWaitGroup(parallelStartupJobs)
	for _, v := range dir {
		loadWg.Add()
		go func(id string) {
			loadWg.Up()
			defer loadWg.Down()
			// ...
		}(v.Name())
	}
	loadWg.Wait()
```

to

```go
	var loadWg sync.WaitGroup
	loadSem := make(chan struct{}, parallelStartupJobs)
	for _, v := range dir {
		loadWg.Add(1)
		go func(id string) {
			loadSem <- struct{}{}
			defer func() {
				<-loadSem
				loadWg.Done()
			}()
			// ...
		}(v.Name())
	}
	loadWg.Wait()
```

Which is the main reason I combined them (the latter seems far easier to mess up on copy-paste and would be repeated quite a bit). But sure, I can do the conversion if you prefer.

cyphar commented:
I've updated the PR to remove the `LimitedWaitGroup` abstraction. Personally I think this version is a bit easier to screw up if we add more code to initialisation (since you need to make sure to do both `<-sem` and `wg.Done()`) but if that's what you'd prefer I'm okay with that too.

cyphar commented:
/ping @cpuguy83

cyphar commented:
There we go.

/cc @cpuguy83

cpuguy83 commented on daemon/daemon_linux.go:
```diff
@@ -14,13 +14,33 @@ import (
+// exceed RLIMIT_NOFILE. If parallelStartupJobs is too large, we reduce it and
+// give a warning (since in theory the user should increase their ulimits to
+// the largest possible value for dockerd).
+func init() {
```
Is this neccessary? In reality we don't know how many containers are running here and seems a bit heavy...

If we do keep this, I think we should call it during daemon init instead of package init.

cpuguy83 replied:
We could also take into account the number of containers, and warn that the limit is low while also adjusting the semaphore size accordingly.

cyphar replied:
I would suggest running Docker with an `NOFILE` limit of `1024` (the systemd default) will certainly cause many other unrelated problems.

But sure, I can make it happen depending on how many containers we actually have.

cyphar replied:
Sorry, another problem with adding it to `daemon/daemon.go` is that Windows doesn't have `ulimit`s -- which is why I've only made the override happen in `daemon_linux.go`.

cyphar replied:
I will also switch to `golang.org/x/sync/semaphore`.

thaJeztah commented:
@unclejack @tonistiigi @kolyshkin PTAL

cyphar commented:
@cpuguy83 Fixed, PTAL.

cpuguy83 reviewed (commented):
LGTM assuming CI says it's good.

cyphar commented:
The failures are just flaky swarm tests.

cyphar commented:
The failures are flaky swarm tests.

/ping @cpuguy83

olljanat commented:
Derek add label: rebuild/*

olljanat commented:
Derek add label: rebuild/*

olljanat commented:
Derek remove label: rebuild/*

cyphar commented on daemon/daemon.go:
```diff
@@ -207,68 +210,108 @@ func (daemon *Daemon) restore() error {
-			if err := daemon.mergeAndVerifyLogConfig(&c.HostConfig.LogConfig); err != nil {
-				logrus.Errorf("Failed to verify log config for container %s: %q", c.ID, err)
-				continue
+			if err := daemon.registerName(c); err != nil {
```
I'm testing 20K containers that have

```
% docker run --name container$i ubuntu:latest true
```

So no restart logic or live-restore. It's just on-disk state, and it takes 6 minutes to start Docker. I'm sure live-restore will make this slower -- but I think that it's completely unacceptable for Docker to take 6 minutes before the API will respond if you have 20K **completely dead** containers. One minute is still unacceptable, but fixing that will require some very radical changes to how container handling is done in Docker (namely, things should really be lazy-loaded).

thaJeztah commented:
@tonistiigi PTAL; is this good to go with @cyphar's latest information?

cyphar commented on daemon/daemon.go:
```diff
@@ -207,68 +210,108 @@ func (daemon *Daemon) restore() error {
+		go func(id string) {
+			defer group.Done()
+			if err := sem.Acquire(ctx, 1); err != nil {
+				// ctx is done.
```
I will switch to an errgroup.

cyphar replied:
Never mind, switching to an errgroup doesn't make sense and makes the code too complicated. I will use `context.Background`.

cyphar commented:
I've handled the comment by @tonistiigi about `context.Background`. I think that's everything done, but obviously I'd wait for his review.

thaJeztah commented:
Moved this to "merge", thanks @cyphar @tonistiigi !
